// steamid/dist/enums.js
"use strict";
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Instances = exports.Types = exports.Universes = void 0;
var Universes;
(function (Universes) {
    Universes[Universes["INVALID"] = 0] = "INVALID";
    Universes[Universes["PUBLIC"] = 1] = "PUBLIC";
    Universes[Universes["BETA"] = 2] = "BETA";
    Universes[Universes["INTERNAL"] = 3] = "INTERNAL";
    Universes[Universes["DEV"] = 4] = "DEV";
})(Universes = exports.Universes || (exports.Universes = {}));
var Types;
(function (Types) {
    Types[Types["INVALID"] = 0] = "INVALID";
    Types[Types["INDIVIDUAL"] = 1] = "INDIVIDUAL";
    Types[Types["MULTISEAT"] = 2] = "MULTISEAT";
    Types[Types["GAMESERVER"] = 3] = "GAMESERVER";
    Types[Types["ANON_GAMESERVER"] = 4] = "ANON_GAMESERVER";
    Types[Types["PENDING"] = 5] = "PENDING";
    Types[Types["CONTENT_SERVER"] = 6] = "CONTENT_SERVER";
    Types[Types["CLAN"] = 7] = "CLAN";
    Types[Types["CHAT"] = 8] = "CHAT";
    Types[Types["P2P_SUPER_SEEDER"] = 9] = "P2P_SUPER_SEEDER";
    Types[Types["ANON_USER"] = 10] = "ANON_USER";
})(Types = exports.Types || (exports.Types = {}));
var Instances;
(function (Instances) {
    Instances[Instances["ALL"] = 0] = "ALL";
    Instances[Instances["DESKTOP"] = 1] = "DESKTOP";
    Instances[Instances["CONSOLE"] = 2] = "CONSOLE";
    Instances[Instances["WEB"] = 4] = "WEB";
})(Instances = exports.Instances || (exports.Instances = {}));
